lib/dataformat/dbxml/db2xml.3
.TH DB2XML 3 "$Date$"
.SH NAME
db2xml \- convert a database view to XML
.SH SYNOPSIS
.nf
#include "dbxml.h"

int \fBdb2xml\fP(Dbptr \fIdb\fP, char *\fIrootnode\fP, char *\fIrownode\fP,
                 Tbl *\fIfields\fP, Tbl *\fIexpressions\fP, void **\fIxml\fP,
                 int \fIflags\fP)
.fi
.SH DESCRIPTION
\fBdb2xml\fP returns an XML (eXtensible Markup Language) representation of a
database view, returning 0 for success and negative values on error.
The view is entirely contained in a single XML element named after the
view name (table name for a base table), unless an alternative name is
specified with the \fIrootnode\fP argument (set \fIrootnode\fP to null for default
behavior).

Each row of the view is entered as an element named \fIrow\fP unless an
alternative is specified with the \fIrownode\fP argument. If both \fIfields\fP
and \fIexpressions\fP are zero, all \fIfields\fP present in the database view
are used. Each appears as its own element, named the same as the field.
If \fIfields\fP is nonzero and \fIexpressions\fP is zero, the \fIfields\fP specified
are used both as element names and as \fIexpressions\fP (usually simple \fIfields\fP)
from the database. If a table of \fIexpressions\fP is supplied, each expression is
evaluated against each row and the result placed in the corresponding field
name from the \fIfields\fP table.

By default (i.e. with \fIflags\fP set to 0) the \fIxml\fP result is
returned as a string in the \fIxml\fP pointer.  This string
is freshly allocated on each call. Previous contents of the pointer are
ignored. Freeing the returned string is the responsibility of the user. If the
\fIflags\fP argument is set to DBXML_BNS, the returned argument
\fIxml\fP is the address of a pointer to a Buffered Network Stream
object, which allows the XML result to be read out in pieces. For
further detail see bns(3). Once again, freeing the returned Bns structure is 
the responsibility of the user.

.SH EXAMPLE
.ft CW
.in 2c
.nf

.ne 5
yukon% cat try.c
#include <stdlib.h>
#include "db.h"
#include "stock.h"
#include "dbxml.h"

.ne 6
main() {
        Dbptr   db;
        char    *dbname = "/opt/antelope/data/db/demo/demo";
        char    expr[STRSZ];
        char    *xml;
        int     orid;

.ne 7
        dbopen( dbname, "r", &db );

        db = dblookup( db, 0, "origin", 0, 0 );
        db.record = 0;
        dbgetv( db, 0, "orid", &orid, 0 );
        sprintf( expr, "orid == %d", orid );
        db = dbsubset( db, expr, 0 );

.ne 7
        db2xml( db, 0, 0, 0, 0, (void **) &xml, 0 );

        printf( "%s\\n", xml );

        exit( 0 );
}
yukon%

.ne 31
yukon% try
<View45>
   <row>
      <lat>40.0740</lat>
      <lon>69.1640</lon>
      <depth>155.1660</depth>
      <time>704371900.66886</time>
      <orid>1</orid>
      <evid>-1</evid>
      <jdate>1992118</jdate>
      <nass>7</nass>
      <ndef>7</ndef>
      <ndp>-1</ndp>
      <grn>715</grn>
      <srn>48</srn>
      <etype>-</etype>
      <review></review>
      <depdp>-999.0000</depdp>
      <dtype>f</dtype>
      <mb>2.62</mb>
      <mbid>1</mbid>
      <ms>-999.00</ms>
      <msid>-1</msid>
      <ml>-999.00</ml>
      <mlid>-1</mlid>
      <algorithm>locsat:kyrghyz</algorithm>
      <auth>JSPC</auth>
      <commid>-1</commid>
      <lddate>790466871.00000</lddate>
   </row>
</View45>

yukon%

.fi
.in
.ft R
.SH RETURN VALUES
\fBdb2xml\fP returns 0 upon success, -1 upon error. Any errors encountered
are registered with the elog(3) facility.
.SH LIBRARY
-ldbxml $(DBLIBS)
.SH "SEE ALSO"
.nf
db2xml(1), dbselect(3)
.fi
.SH "BUGS AND CAVEATS"
This version has no support for attributes.

The pretty-print indent width is fixed, and newlines are added by default.
At some point it may be necessary to add these later.

The \fIxml\fP pointer is a void pointer to support multiple return
types.  It may be desirable in the future to return something more
complex than a pretty-printed string, such as a libxml2 DOM (Document
Object Model) tree.  If necessary, the returned string can always be
parsed into a representation of choice.

.SH AUTHOR
.nf
Kent Lindquist
Lindquist Consulting
.fi
.\" $Id$


lib/dataformat/dbxml/db2xml.c
/* 
 * db2xml.c
 * Kent Lindquist
 * Lindquist Consulting
 * 2003-2004
 */

#include <stdio.h>
#include <string.h>
#include "db.h"
#include "stock.h"
#include "coords.h"
#include "bns.h"
#include "dbxml.h"

static void 
safe_strsub( char *source, char *pattern, char *replacement, char *dest )
{
	char 	*copy;

	allot( char *, copy, strlen( source ) + STRSZ );

	strsub( source, pattern, replacement, copy );
	
	strcpy( dest, copy );

	free( copy );
}

static void
add_starttag( void **vstack, char *tagname )
{
	char	*copy;

	copy = strdup( tagname );
	strtrim( copy );

	pushstr( vstack, "<" );
	pushstr( vstack, copy );
	pushstr( vstack, ">" );

	free( copy );
}

static void
add_endtag( void **vstack, char *tagname )
{
	char	*copy;

	copy = strdup( tagname );
	strtrim( copy );

	pushstr( vstack, "</" );
	pushstr( vstack, copy );
	pushstr( vstack, ">" );

	free( copy );
}

static void
add_dataelement( void **vstack, char *tagname, char *value )
{
	char	*copy;

	copy = strdup( value );
	strtrim( copy );

	safe_strsub( copy, "&", "&amp;", copy );
	safe_strsub( copy, "<", "&lt;", copy );
	safe_strsub( copy, ">", "&gt;", copy );

	add_starttag( vstack, tagname );
	pushstr( vstack, copy );
	add_endtag( vstack, tagname );

	free( copy );
}

static void
add_fieldname( Tbl *fieldnames, char *tablename, char *fieldname )
{
	char	qualified[STRSZ];

	sprintf( qualified, "%s.%s", tablename, fieldname );

	pushtbl( fieldnames, strdup( qualified ) );

	return;
}

static Tbl *
get_fieldnames( Dbptr db, int flags )
{
	Tbl	*fieldnames;
	Tbl	*view_tables;
	Tbl	*table_fields;
	char	*tablename;
	char	*fieldname;
	char	afield[STRSZ];
	Dbvalue	result;
	int	itable;
	int	ifield;
	static Hook *hook = 0;
	int	start;
	int	nchars;
	int	is_view;

	fieldnames = newtbl( 0 );
	view_tables = newtbl( 0 );

	dbquery( db, dbTABLE_IS_VIEW, &is_view );

	if( is_view ) {

		dbquery( db, dbVIEW_TABLES, &view_tables );

		view_tables = duptbl( view_tables, strdup );

	} else {

		dbquery( db, dbTABLE_NAME, &result );

		view_tables = strtbl( strdup( result.t ), 0 );
	}

	for( itable = 0; itable < maxtbl( view_tables ); itable++ ) {

		tablename = gettbl( view_tables, itable );

		db = dblookup( db, "", tablename, "", "" );

		if( flags & DBXML_PRIMARY ) {

			dbquery( db, dbPRIMARY_KEY, &table_fields );

		} else {

			dbquery( db, dbTABLE_FIELDS, &table_fields );
		}

		for( ifield = 0; ifield < maxtbl( table_fields ); ifield++ ) {
			
			fieldname = gettbl( table_fields, ifield );

			if( strcontains( fieldname, "::", &hook,
					 &start, &nchars ) ) {
				
				strcpy( afield, fieldname );
				afield[start] = '\0';
				fprintf( stderr, "SCAFFOLD: adding %s\n", afield);
				add_fieldname( fieldnames, tablename, afield );

				strcpy( afield, fieldname + start + 2 );
				fprintf( stderr, "SCAFFOLD: adding %s\n", afield);
				add_fieldname( fieldnames, tablename, afield );

			} else {

				add_fieldname( fieldnames, tablename, fieldname );
			}
		}
	}

	freetbl( view_tables, free );

	return fieldnames;
}

int
db2xml( db, rootnode, rownode, fields_in, expressions_in, xml, flags )
Dbptr 	db; 
char 	*rootnode;
char 	*rownode;
Tbl 	*fields_in;
Tbl 	*expressions_in;
void 	**xml;
int 	flags;
{
	Bns	*xml_bns = 0;
	char	*xmlstring = 0;
	char	*field; 
	char	*fieldname; 
	char	*vstack = 0;
	Tbl	*fields = 0;
	Tbl	*expressions = 0;
	int	i;
	int	n;
	int	setflag = 0; 
	char	*separator = "\n";
	char	*indent = "   ";
	char	*roottag = 0;
	char	*rowtag = 0;
	char	*default_rowtag = "row";
	Expression **expr; 
	Dbvalue	result;
	char	temp[STRSZ]; 
	int 	retcode = 0; 
	int	ns, ne; 
	int	free_fieldnames = 0;

	if( db.table < 0 ) {
		register_error( 0, "db2xml: not a view or a table\n" );
		return -1;
	}

	if( ( flags & DBXML_PRIMARY ) && 
	    ( fields_in != 0 || expressions_in != 0 ) ) {

		register_error( 0, "db2xml: fields are explicitly specified; "
				   "ignoring useless request for primary keys\n" );
	}

	if( rootnode != 0 && *rootnode != 0 ) {
		
		roottag = strdup( rootnode );

	} else {
		
		dbquery( db, dbTABLE_NAME, &result );
		roottag = strdup( result.t );
	} 
	add_starttag( (void **) &vstack, roottag );
	pushstr( (void **) &vstack, separator );

	if( rownode != 0 && *rownode != 0 ) {
		
		rowtag = strdup( rownode );

	} else {
		
		rowtag = strdup( default_rowtag );
	} 

	if( fields_in == 0 && expressions_in == 0 ) {

		fields = expressions = get_fieldnames( db, flags );

		free_fieldnames++;


	} else if( expressions_in == 0 ) {

		expressions = fields = fields_in;

	} else if( fields_in == 0 ) {

		register_error( 0, 
		"db2xml: must specify field names with nonzero list "
		"of expressions\n" );
		return -1;

	} else {

		fields = fields_in;
		expressions = expressions_in;
	}

	if( maxtbl( fields ) != maxtbl( expressions ) ) {

		register_error( 0, 
		"db2xml: number of fields must match number of expressions\n" );
		return -1;
	}

	dbget_range ( db, &ns, &ne ); 

	n = maxtbl( expressions );
	allot ( Expression **, expr, n ); 
	for (i = 0; i < n; i++)
	{
		field = (char *) gettbl( expressions , i);
		if ( dbex_compile ( db, field, &(expr[i]), 0 ) < 0 ) {
		  	return -1;
		}
	}

	for (db.record = ns; db.record < ne; db.record++ ) {

		pushstr( (void **) &vstack, indent );
		add_starttag( (void **) &vstack, rowtag );
		pushstr( (void **) &vstack, separator );

		for ( i=0; i<n; i++ ) {
			fieldname = (char *) gettbl( fields, i);
			if (dbex_eval ( db, expr[i], setflag, &result ) < 0 ) {
				retcode = -1;
				sprintf( temp, "***Bad Expression***"); 
			} else {
				switch ( expr[i]->type ) {
				case dbBOOLEAN: 
			    		sprintf ( temp, "%-5s", result.i ? "true" : "false" ); 
			    		break;
				case dbYEARDAY:
				case dbINTEGER:  
			    		sprintf ( temp, "%8d", result.i ); 
			    		break;
				case dbTIME:
			    		sprintf ( temp, "%17.5lf", result.d ); 
			    		break; 
				case dbREAL:
			    		sprintf ( temp, "%15.8g", result.d ); 
			    		break;
				case dbSTRING:
			    		sprintf ( temp, "%-10s", result.t ); 
			    		free(result.t);
			    		result.t = 0;
			    		break;
				default:
			    		sprintf ( temp, "**INVALID**" ); 
			    		break;
				}
			}

			if ( indent != 0 ) {

				pushstr( (void **) &vstack, indent );
				pushstr( (void **) &vstack, indent );
			}

			add_dataelement( (void **) &vstack, fieldname, temp );

			if ( separator != 0 ) {

				pushstr( (void **) &vstack, separator );
			}
		}

		pushstr( (void **) &vstack, indent );
		add_endtag( (void **) &vstack, rowtag );
		pushstr( (void **) &vstack, separator );
	}

	if( free_fieldnames ) {

		freetbl( fields, free );
	}

	for ( i=0; i<n; i++ ) {
		dbex_free ( expr[i] ); 
	}

	free(expr);

	add_endtag( (void **) &vstack, roottag );
	pushstr( (void **) &vstack, separator );

	free( roottag );
	free( rowtag );

	xmlstring = popstr( (void **) &vstack, 1 );

	if( flags & DBXML_BNS ) {

		xml_bns = bnsnew( -1, strlen( xmlstring ) + 2 );

		buf2bns( xml_bns, xmlstring, strlen( xmlstring ) + 1 );

		free( xmlstring );

		*xml = (void *) xml_bns;

	} else {
		
		*xml = (void *) xmlstring;
	}

	return retcode;
}
